mlir/lib/Dialect/SPIRV/Transforms/SPIRVWebGPUTransforms.cpp

```diff
@@ -133,6 +133,48 @@ Value lowerExtendedMultiplication(Operation *mulOp, PatternRewriter &rewriter,
       loc, mulOp->getResultTypes().front(), llvm::ArrayRef({low, high}));
 }
 
+Value lowerCarryAddition(Operation *addOp, PatternRewriter &rewriter, Value lhs,
+                         Value rhs) {
+  Location loc = addOp->getLoc();
+  Type argTy = lhs.getType();
+  // Emulate 64-bit addition by splitting each input element of type i32 to
+  // i16 similar to above in lowerExtendedMultiplication. We then expand
+  // to 3 additions:
+  //    - Add two low digits into low resut
+  //    - Add two high digits into high result
+  //    - Add the carry from low result to high result
```

Comment:
Could we calculate the carry result by comparing the sum to one of the operands?
> Member 1 of the result gets the high-order (carry) bit of the result of the addition. That is, it gets the value 1 if the addition overflowed the component width, and 0 otherwise.

I think it's something like:
```
carry := (a + b) < a ? 1 : 0
```

Where `<` is unsigned less than.

This way we wouldn't have to split the values at all.

Reply:
Nice, that feels much better. From now reading the linked spec a second time, it is clear to me that we can expect the overflow behaviour to be how we would want. I was not diligent enough when reading the "avoid overflow and underflow" statement the first time, whoops.

https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_arithmetic_instructions